cl_dll/tri.cpp

```diff
@@ -40,6 +60,91 @@ void CL_DLLEXPORT HUD_DrawNormalTriangles( void )
 void RunEventList( void );
 #endif
 
+void DivideRGBABy255(float &r, float &g, float &b, float &a)
+{
+	r /= 255.0f;
+	g /= 255.0f;
+	b /= 255.0f;
+	a /= 255.0f;
+}
+
+void DrawAACuboid(triangleapi_s *pTriAPI, Vector corner1, Vector corner2)
+{
+	pTriAPI->Begin(TRI_QUADS);
+
+	pTriAPI->Vertex3f(corner1.x, corner1.y, corner1.z);
+	pTriAPI->Vertex3f(corner1.x, corner2.y, corner1.z);
+	pTriAPI->Vertex3f(corner2.x, corner2.y, corner1.z);
+	pTriAPI->Vertex3f(corner2.x, corner1.y, corner1.z);
+
+	pTriAPI->Vertex3f(corner1.x, corner1.y, corner1.z);
+	pTriAPI->Vertex3f(corner1.x, corner1.y, corner2.z);
+	pTriAPI->Vertex3f(corner1.x, corner2.y, corner2.z);
+	pTriAPI->Vertex3f(corner1.x, corner2.y, corner1.z);
+
+	pTriAPI->Vertex3f(corner1.x, corner1.y, corner1.z);
+	pTriAPI->Vertex3f(corner2.x, corner1.y, corner1.z);
+	pTriAPI->Vertex3f(corner2.x, corner1.y, corner2.z);
+	pTriAPI->Vertex3f(corner1.x, corner1.y, corner2.z);
+
+	pTriAPI->Vertex3f(corner2.x, corner2.y, corner2.z);
+	pTriAPI->Vertex3f(corner1.x, corner2.y, corner2.z);
+	pTriAPI->Vertex3f(corner1.x, corner1.y, corner2.z);
+	pTriAPI->Vertex3f(corner2.x, corner1.y, corner2.z);
+
+	pTriAPI->Vertex3f(corner2.x, corner2.y, corner2.z);
+	pTriAPI->Vertex3f(corner2.x, corner1.y, corner2.z);
+	pTriAPI->Vertex3f(corner2.x, corner1.y, corner1.z);
+	pTriAPI->Vertex3f(corner2.x, corner2.y, corner1.z);
+
+	pTriAPI->Vertex3f(corner2.x, corner2.y, corner2.z);
+	pTriAPI->Vertex3f(corner2.x, corner2.y, corner1.z);
+	pTriAPI->Vertex3f(corner1.x, corner2.y, corner1.z);
+	pTriAPI->Vertex3f(corner1.x, corner2.y, corner2.z);
+
+	pTriAPI->End();
+}
+
+void DrawServerTriggers()
+{
+	if ((gHUD.m_pShowServerTriggers->value > 0) && (gHUD.m_pShowServerTriggers->value != 2.0f))
+	{
+		for (int e = 0; e < MAX_EDICTS; ++e)
+		{
+			cl_entity_t* ent = gEngfuncs.GetEntityByIndex(e);
+			if (ent)
+			{
+				if (ent->model)
+				{
+					if ((ent->curstate.rendermode == kRenderTransColor) && (ent->curstate.renderfx == kRenderFxTrigger))
+					{
+						color24 colors = ent->curstate.rendercolor;
+						if (!gHUD.IsTriggerForSinglePlayer(colors))
+						{
+							gEngfuncs.pTriAPI->RenderMode(kRenderTransAdd);
+							gEngfuncs.pTriAPI->CullFace(TRI_NONE);
+
+							float r = colors.r, g = colors.g, b = colors.b, a = std::min(255.0f, std::max(0.0f, gHUD.m_pShowServerTriggersAlpha->value));
+							DivideRGBABy255(r, g, b, a);
+							gEngfuncs.pTriAPI->Color4f(r, g, b, a);
```

Comment:
Maybe `Color4ub()` then you don't need to divide?

Reply:
Month ago, I had to reverse-engineering all TriAPI functions from hardware engine (since I am preparing for a major refactor update for BXT) and I can tell you that using `pTriAPI->Color4ub()` will not be a good idea, take a look yourself:

```cpp
void tri_GL_Color4ub(unsigned char r, unsigned char g, unsigned char b, unsigned char a) // pTriAPI->Color4ub
{
	gGlR = r * (1.0f / 255.0f);
	gGlG = g * (1.0f / 255.0f);
	gGlB = b * (1.0f / 255.0f);
	gGlW = a * (1.0f / 255.0f);
	glColor4f(gGlR, gGlG, gGlB, 1.0f);
}
```

Yeah, it divides all of that colors to 255, but as can be seen, `glColor4f` is would be still hardcoded to 1

If someone doubt my skills, well, open hardware linux engine in IDA Pro or Ghidra and you will see the same:

![color4ub](https://github.com/YaLTeR/OpenAG/assets/58108407/620742af-1dce-4977-ba26-8132f95c6720)

Reply:
amazing